feat(forms): add updateOn blur option to FormControls

kara opened the pull request:
By default, the value and validation status of a `FormControl` updates whenever its value changes. If an application has heavy validation requirements, updating on every text change can sometimes be too expensive. 

This commit introduces a new option that improves performance by delaying form control updates until the "blur" event.  To use it, set the `updateOn` option to `blur` when instantiating the `FormControl`.

```ts
// example without validators
const c = new FormControl('', { updateOn: 'blur' });

// example with validators
const c= new FormControl('', {
   validators: Validators.required,
   updateOn: 'blur'
});
```

Like in AngularJS, setting `updateOn` to `blur` will delay the update of the value as well as the validation status. Updating value and validity together keeps the system easy to reason about, as the two will always be in sync.  It's also worth noting that the value/validation pipeline does still run when the form is initialized (in order to support initial values).

**This is just the first PR**. Upcoming PRs will address:

- Other `updateOn` values, like `submit`
- Setting `updateOn` at a form-wide or application-wide level
- Support in template-driven forms
- Option for skipping initial validation run or more global error display configuration
- Better support of reactive validation strategies

See more context in [design doc](https://docs.google.com/document/d/1dlJjRXYeuHRygryK0XoFrZNqW86jH4wobftCFyYa1PA/edit#heading=h.r6gn0i8f19wz).

mary-poppins commented:
You can preview 1e1ff76 at https://pr18408-1e1ff76.ngbuilds.io/.

mary-poppins commented:
You can preview 8296d32 at https://pr18408-8296d32.ngbuilds.io/.

mary-poppins commented:
You can preview c3c692e at https://pr18408-c3c692e.ngbuilds.io/.

tinayuangao commented on packages/forms/src/model.ts:
```diff
@@ -78,11 +78,15 @@ function coerceToAsyncValidator(
                                              origAsyncValidator || null;
 }
 
+export type FormHooks = 'change' | 'blur';
```
nit: do we want to add some docs for this?

kara replied:
Normally yes, but this isn't actually exported by the forms module as part of the public API. This is just exported so the type can be used elsewhere internally.

kara replied:
Re docs in general, I'm working on another PR to update the form validation guide with an extended section on this.

tinayuangao commented on packages/forms/test/form_control_spec.ts:
```diff
@@ -76,7 +76,22 @@ export function main() {
+      it('should default to onChange with an options obj', () => {
+        const c = new FormControl('', {validators: Validators.required});
+        expect(c._updateOn).toEqual('change');
+      });
```
Also test when `{updateOn: 'blur'}`, `c._updateOn`'s value is correct?

kara replied:
Good call; will add a test.

tinayuangao reviewed (approved):
LGTM

kara committed:
feat(forms): add updateOn blur option to FormControls

By default, the value and validation status of a `FormControl` updates
whenever its value changes. If an application has heavy validation
requirements, updating on every text change can sometimes be too expensive.

This commit introduces a new option that improves performance by delaying
form control updates until the "blur" event.  To use it, set the `updateOn`
option to `blur` when instantiating the `FormControl`.

```ts
// example without validators
const c = new FormControl(, { updateOn: blur });

// example with validators
const c= new FormControl(, {
   validators: Validators.required,
   updateOn: blur
});
```

Like in AngularJS, setting `updateOn` to `blur` will delay the update of
the value as well as the validation status. Updating value and validity
together keeps the system easy to reason about, as the two will always be
in sync. It's  also worth noting that the value/validation pipeline does
still run when the form is initialized (in order to support initial values).

Closes #7113

mary-poppins commented:
You can preview dc62ecb at https://pr18408-dc62ecb.ngbuilds.io/.

cexbrayat commented:
@kara Looks like we can pass the `updateOn: 'blur'` option to a `FormGroup` too, but it doesn't have any effect? It may be a bug, or that shouldn't be possible as it's misleading. Probably the same with `FormArray`.

Also, is it on purpose that the `control` (and `group`, and `array`) methods of the `FormBuilder` don't accept `AbstractControlOptions`?

kara commented:
@cexbrayat This feature is still in development; this is just the first PR.  You may want to check out the PR desc and its linked design doc for more context for what's coming next.

cexbrayat commented:
@kara Thank you for your answer. I did read the design doc and did not see the answer to these question, that's why I dared to ask (but maybe I missed it). Great if it comes in a following PR, I'm looking forward to try it!

tyteen4a03 commented:
Do I have to specify this for every sub-FormGroups?
